contracts/Staking.sol

```diff
@@ -0,0 +1,574 @@
+      (resCode, ackPackage) = _handleDistributeUndelegatedSynPackage(iter);
+    } else {
+      require(false, "unknown event type");
+    }
+
+    if (resCode != CODE_OK) {
+      emit failedSynPackage(eventType, resCode);
+    }
+    return ackPackage;
+  }
+
+  function handleAckPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract initParams override {
+    RLPDecode.Iterator memory iter = msgBytes.toRLPItem().iterator();
+    uint8 eventType = uint8(iter.next().toUint());
+    if (eventType == EVENT_DELEGATE) {
+      _handleDelegateAckPackage(iter);
+    } else if (eventType == EVENT_UNDELEGATE) {
+      _handleUndelegateAckPackage(iter);
+    } else if (eventType == EVENT_REDELEGATE) {
+      _handleRedelegateAckPackage(iter);
+    } else {
+      require(false, "unknown event type");
+    }
+    return;
+  }
+
+  function handleFailAckPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract initParams override {
+    RLPDecode.Iterator memory iter = msgBytes.toRLPItem().iterator();
+    uint8 eventType = uint8(iter.next().toUint());
+    if (eventType == EVENT_DELEGATE) {
+      _handleDelegateFailAckPackage(iter);
+    } else if (eventType == EVENT_UNDELEGATE) {
+      _handleUndelegateFailAckPackage(iter);
+    } else if (eventType == EVENT_REDELEGATE) {
+      _handleRedelegateFailAckPackage(iter);
+    } else {
+      require(false, "unknown event type");
+    }
+    return;
+  }
+
+  /***************************** External functions *****************************/
+  function delegate(address validator, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(amount >= minDelegation, "invalid delegate amount");
+    require(msg.value >= amount.add(oracleRelayerFee), "not enough msg value");
+
+    // native bnb decimals is 8 on BBC, while the native bnb decimals on BSC is 18
+    uint256 convertedAmount = amount.div(TEN_DECIMALS);
+    uint256 _oracleRelayerFee = (msg.value).sub(amount);
+
+    bytes[] memory elements = new bytes[](3);
+    elements[0] = msg.sender.encodeAddress();
+    elements[1] = validator.encodeAddress();
+    elements[2] = convertedAmount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(CROSS_STAKE_CHANNELID, _RLPEncode(EVENT_DELEGATE, msgBytes), _oracleRelayerFee.div(TEN_DECIMALS));
+    payable(TOKEN_HUB_ADDR).transfer(msg.value);
+
+    delegated[msg.sender] = delegated[msg.sender].add(amount);
+    delegatedOfValidator[msg.sender][validator] = delegatedOfValidator[msg.sender][validator].add(amount);
+
+    emit delegateSubmitted(msg.sender, validator, amount, _oracleRelayerFee);
+  }
+
+  function undelegate(address validator, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(block.timestamp >= pendingUndelegateTime[msg.sender][validator], "pending undelegation exist");
+    if (amount < minDelegation) {
```

Comment:
Is it necessary to check that the current delegator has enough amount to undelegate？